### PR TITLE
MSI-2558: Introduce reusable checker classes

### PR DESCRIPTION
**Pull request by vadimjustus**
### Description (*)
Introduce validation checker classes in order to reuse checks instead of copying them between different validators.

### Fixed Issues (if relevant)
1. magento/inventory#2558: Source validator duplicate logic

### Manual testing scenarios (*)
1. Run integration tests
2. Create new sources/stocks via Magento Admin UI
3. Create new sources/stocks via WebAPI

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: app/code/Magento/Inventory/Model/Source/Validator/CodeValidator.php [hunk 1]
```diff
@@ -37,15 +65,12 @@ public function validate(SourceInterface $source): ValidationResult
...
-            $errors[] = __('"%field" can not contain whitespaces.', ['field' => SourceInterface::SOURCE_CODE]);
-        } elseif (preg_match('/\$[:]*{(.)*}/', $value)) {
-            $errors[] = __('Validation Failed');
-        } else {
-            $errors = [];
-        }
+        $errors = [];
+        $errors[] = $this->notAnEmptyString->execute(SourceInterface::SOURCE_CODE, $value);
+        $errors[] = $this->noWhitespaceInString->execute(SourceInterface::SOURCE_CODE, $value);
+        $errors[] = $this->noSpecialCharsInString->execute($value);
```
Comment (Stepa4man) [inline app/code/Magento/Inventory/Model/Source/Validator/CodeValidator.php]: What do you think about:
```
$errors = [
        $this->notAnEmptyString->execute(SourceInterface::SOURCE_CODE, $value),
        $this->noWhitespaceInString->execute(SourceInterface::SOURCE_CODE, $value),
        $this->noSpecialCharsInString->execute($value)
];
```

File: app/code/Magento/Inventory/Model/Source/Validator/NameValidator.php [hunk 2]
```diff
@@ -37,13 +56,11 @@ public function validate(SourceInterface $source): ValidationResult
...
-        if ('' === trim($value)) {
-            $errors[] = __('"%field" can not be empty.', ['field' => SourceInterface::NAME]);
-        } elseif (preg_match('/\$[:]*{(.)*}/', $value)) {
-            $errors[] = __('Validation Failed');
-        } else {
-            $errors = [];
-        }
+        $errors = [];
+        $errors[] = $this->notAnEmptyString->execute(SourceInterface::NAME, $value);
+        $errors[] = $this->noSpecialCharsInString->execute($value);
```
Comment (Stepa4man) [inline app/code/Magento/Inventory/Model/Source/Validator/NameValidator.php]: Same here

File: app/code/Magento/Inventory/Model/SourceItem/Validator/SkuValidator.php [hunk 3]
```diff
@@ -22,26 +23,33 @@ class SkuValidator implements SourceItemValidatorInterface
...
         $this->validationResultFactory = $validationResultFactory;
+        $this->notAnEmptyString = $notAnEmptyString;
     }
 
     /**
      * @inheritdoc
      */
     public function validate(SourceItemInterface $source): ValidationResult
     {
+        $value = (string)$source->getSku();
```
Comment (Stepa4man) [inline app/code/Magento/Inventory/Model/SourceItem/Validator/SkuValidator.php]: You don't have to cast SKU.

File: app/code/Magento/Inventory/Model/SourceItem/Validator/SourceCodeValidator.php [hunk 4]
```diff
@@ -37,13 +57,11 @@ public function validate(SourceItemInterface $source): ValidationResult
...
-        if ('' === trim($value)) {
-            $errors[] = __('"%field" can not be empty.', ['field' => SourceItemInterface::SOURCE_CODE]);
-        } elseif (preg_match('/\s/', $value)) {
-            $errors[] = __('"%field" can not contain whitespaces.', ['field' => SourceItemInterface::SOURCE_CODE]);
-        } else {
-            $errors = [];
-        }
+        $errors = [];
+        $errors[] = $this->notAnEmptyString->execute(SourceItemInterface::SOURCE_CODE, $value);
+        $errors[] = $this->noWhitespaceInString->execute(SourceItemInterface::SOURCE_CODE, $value);
```
Comment (Stepa4man) [inline app/code/Magento/Inventory/Model/SourceItem/Validator/SourceCodeValidator.php]: Is it better to gather to one declaration?

File: app/code/Magento/Inventory/Model/ValidationChecker/IsNumericValue.php [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+/**
+ * Checks whether given string is empty
+ */
+class IsNumericValue
+{
+    /**
+     * Checks whether given string is empty
+     *
+     * @param string $fieldName
+     * @param mixed $value
```
Comment (Stepa4man) [inline app/code/Magento/Inventory/Model/ValidationChecker/IsNumericValue.php]: Can we make it not a mixed?

File: app/code/Magento/Inventory/Model/ValidationChecker/NoSpecialCharsInString.php [hunk 6]
```diff
@@ -0,0 +1,31 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\Inventory\Model\ValidationChecker;
```
Comment (Stepa4man) [inline app/code/Magento/Inventory/Model/ValidationChecker/NoSpecialCharsInString.php]: Validation checker... Sounds like "weapon shooter". Is it possible to give it more appropriate namespace?

File: app/code/Magento/Inventory/Model/ValidationChecker/NoSpecialCharsInString.php [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+/**
+ * Checks whether given string contains special chars
+ */
+class NoSpecialCharsInString
+{
+    /**
+     * Checks whether given string contains special chars
+     *
+     * @param mixed $value
```
Comment (Stepa4man) [inline app/code/Magento/Inventory/Model/ValidationChecker/NoSpecialCharsInString.php]: Can we avoid using mixed?

File: app/code/Magento/Inventory/Model/ValidationChecker/NoWhitespaceInString.php [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+/**
+ * Checks whether given value contains whitespace
+ */
+class NoWhitespaceInString
+{
+    /**
+     * Checks whether given value contains whitespace
+     *
+     * @param string $fieldName
+     * @param mixed $value
```
Comment (Stepa4man) [inline app/code/Magento/Inventory/Model/ValidationChecker/NoWhitespaceInString.php]: Same with mixed here

**Review by Stepa4man — DISMISSED** [review]
Hi @vadimjustus 
Thanks for your contribution!

Please check my comments

**Review by sidolov — DISMISSED** [review]

**Comment by sidolov** [comment]
@magento run Unit Tests

**Review by sidolov — APPROVED** [review]

**Pull request merged**